File: backend/security.py
import hashlib
import hmac
import secrets

PBKDF2_ITERATIONS = 390_000


def hash_password(password: str) -> str:
    salt = secrets.token_hex(16)
    digest = hashlib.pbkdf2_hmac(
        "sha256",
        password.encode("utf-8"),
        bytes.fromhex(salt),
        PBKDF2_ITERATIONS,
    ).hex()
    return f"pbkdf2_sha256${PBKDF2_ITERATIONS}${salt}${digest}"


def verify_password(password: str, stored: str | None) -> bool:
    if not stored:
        return False

    try:
        algorithm, iterations, salt, digest = stored.split("$")
    except ValueError:
        return False

    if algorithm != "pbkdf2_sha256":
        return False

    candidate = hashlib.pbkdf2_hmac(
        "sha256",
        password.encode("utf-8"),
        bytes.fromhex(salt),
        int(iterations),
    ).hex()

    return hmac.compare_digest(candidate, digest)


def generate_token() -> str:
    return secrets.token_urlsafe(32)


def hash_token(token: str) -> str:
    return hashlib.sha256(token.encode("utf-8")).hexdigest()


PASSWORD_RULE_MESSAGE = (
    "Password must be at least 8 characters and contain "
    "at least one letter and one number."
)


def validate_password(password: str) -> str | None:
    """Return None when valid, otherwise a human-readable reason."""
    if len(password) < 8:
        return PASSWORD_RULE_MESSAGE
    if not any(char.isalpha() for char in password):
        return PASSWORD_RULE_MESSAGE
    if not any(char.isdigit() for char in password):
        return PASSWORD_RULE_MESSAGE
    return None

File: lib/security.ts
import { pbkdf2Sync, randomBytes, createHash, timingSafeEqual } from "crypto";

const ITERATIONS = 390_000;

export function hashPassword(password: string): string {
  const salt = randomBytes(16).toString("hex");
  const digest = pbkdf2Sync(
    password,
    Buffer.from(salt, "hex"),
    ITERATIONS,
    32,
    "sha256"
  ).toString("hex");
  return `pbkdf2_sha256$${ITERATIONS}$${salt}$${digest}`;
}

export function verifyPassword(password: string, stored: string | null): boolean {
  if (!stored) return false;

  const parts = stored.split("$");
  if (parts.length !== 4) return false;

  const [algorithm, iterations, salt, digest] = parts;
  if (algorithm !== "pbkdf2_sha256") return false;

  const candidate = pbkdf2Sync(
    password,
    Buffer.from(salt, "hex"),
    parseInt(iterations, 10),
    32,
    "sha256"
  ).toString("hex");

  try {
    return timingSafeEqual(
      Buffer.from(candidate, "hex"),
      Buffer.from(digest, "hex")
    );
  } catch {
    return false;
  }
}

export function generateToken(): string {
  return randomBytes(24).toString("base64url");
}

export function hashToken(token: string): string {
  return createHash("sha256").update(token).digest("hex");
}

const PASSWORD_RULE_MESSAGE =
  "Password must be at least 8 characters and contain at least one letter and one number.";

export function validatePassword(password: string): string | null {
  if (password.length < 8) return PASSWORD_RULE_MESSAGE;
  if (!/[a-zA-Z]/.test(password)) return PASSWORD_RULE_MESSAGE;
  if (!/[0-9]/.test(password)) return PASSWORD_RULE_MESSAGE;
  return null;
}
